src/wp-includes/class-wp-scripts.php

```diff
@@ -714,6 +781,262 @@ public function in_default_dir( $src ) {
+	 *
+	 * @since 6.3.0
+	 *
+	 * @param string $handle   The script handle.
+	 * @param string $position Position of the inline script.
+	 *
+	 * @return bool True if script present. False if empty.
+	 */
+	private function has_non_standalone_inline_script( $handle, $position ) {
+		$non_standalone_script_key = $position . '-non-standalone';
+		$non_standalone_script     = $this->get_data( $handle, $non_standalone_script_key );
+
+		return ! empty( $non_standalone_script );
+	}
+
+	/**
+	 * Check if all of a scripts dependents are deferrable, which is required to maintain execution order.
+	 *
+	 * @since 6.3.0
+	 *
+	 * @param string $handle  The script handle.
+	 * @param array  $checked Optional. An array of already checked script handles, used to avoid recursive loops.
+	 * @return bool True if all dependents are deferrable, false otherwise.
+	 */
+	private function has_only_deferrable_dependents( $handle, $checked = array() ) {
+		// If this node was already checked, this script can be deferred and the branch ends.
+		if ( in_array( $handle, $checked, true ) ) {
+			return true;
+		}
+
+		$checked[]  = $handle;
+		$dependents = $this->get_dependents( $handle );
+
+		// If there are no dependents remaining to consider, the script can be deferred.
+		if ( empty( $dependents ) ) {
+			return true;
+		}
+
+		// Consider each dependent and check if it is deferrable.
+		foreach ( $dependents as $dependent ) {
+			// If the dependent script is not using the defer or async strategy, no script in the chain is deferrable.
+			$strategy = $this->get_intended_strategy( $dependent );
+			if ( ! $this->is_non_blocking_strategy( $strategy ) ) {
+				return false;
+			}
+
+			// If the dependent script has a non-standalone inline script in the 'before' position associated with it, do not defer.
+			if ( $this->has_non_standalone_inline_script( $dependent, 'before' ) ) {
+				return false;
+			}
+
+			// Recursively check all dependents.
+			if ( ! $this->has_only_deferrable_dependents( $dependent, $checked ) ) {
+				return false;
+			}
+		}
+
+		return true;
+	}
+
+	/**
+	 * Get the best eligible loading strategy for a script.
+	 *
+	 * @since 6.3.0
+	 *
+	 * @param string  $handle The registered handle of the script.
+	 * @return string $strategy return the final strategy.
+	 */
+	private function get_eligible_loading_strategy( $handle ) {
+		if ( ! isset( $this->registered[ $handle ] ) ) {
+			return '';
+		}
+
+		$intended_strategy = $this->get_intended_strategy( $handle );
+
+		/*
+		 * Handle known blocking strategy scenarios.
+		 *
+		 * blocking if script args not set.
+		 * blocking if explicitly set.
+		 */
+		if ( '' === $intended_strategy || 'blocking' === $intended_strategy ) {
+			return '';
+		}
+
+		// Handling async strategy scenarios.
+		if ( 'async' === $intended_strategy && empty( $this->registered[ $handle ]->deps ) && empty( $this->get_dependents( $handle ) ) ) {
```

Comment:
I don't think this is necessarily quite right. As I understand, this is saying that the `async` strategy is only available if it has no dependencies and nothing else depends on it. However, asynchronous scripts could very well depend on other scripts. They just have to wait for the other scripts to be loaded. In AMP, for example, all scripts are loaded asynchronously. And yet, component scripts, such as `amp-carousel` depend on the AMP runtime (`v0.js`) to be able to execute.

For example, each runtime-dependent component script contains code like this:

```js
(self.AMP = self.AMP || []).push({ f: () => { /*...*/ } });
```

When the runtime script loads, it checks of `self.AMP` has been defined, and if so, it loops over the items in the array and executes them. It then overrides `self.AMP.push` so that any subsequent calls will result in the component being run right away. In this way, all of AMP's component scripts can be added to the page in any order with all being loaded asynchronously downloaded from the network in any order, and it'll all just work.

I've put together a standalone demonstration of this, which includes showing how an async script can have a dependency on not only a runtime script but also any number of other async scripts: https://async-library-script-loading-demo.glitch.me/

Another example showing async scripts is this [AMP experiment opt-in code](https://github.com/ampproject/amphtml/blob/5e16bc4114cbd05fff74516d39e14cb1183ac8a1/tools/experiments/README.md#enable-automatically-for-invalid-documents-demos-and-automated-tests): 

```html
<script>
  (self.AMP = self.AMP || []).push(function (AMP) {
    AMP.toggleExperiment('my-experiment', true);
  });
</script>
```

This pattern is also commonly seen in the [Async Tracking code for Google Analytics](https://stackoverflow.com/questions/2538252/what-is-var-gaq-gaq-for):

```js
var _gaq = _gaq || []; 
_gaq.push(['_setAccount', 'UA-XXXXX-X']); 
_gaq.push(['_trackPageview']); 

(function() { 
  var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true; 
  ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js'; 
  var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s); 
})(); 
```

All of this to say, just because a script has the `async` strategy should not exclude it from being able to have dependencies. Let's say that `baz` depends on `foo` and `bar`, and all three are asynchronous. It would be handy to be able to just do:

```php
wp_enqueue_script( 'baz' );
```

And have the dependency chain ensure the asynchronous dependencies are automatically printed. Otherwise, for such an async library you would have to manually enqueue all three separately with no dependencies declared on them in order to get them all printed asynchronously:

```php
// Note these can be enqueued in any order since if it's
// an async library, they'll execute in the required order
// even if they are not loaded in the required order.
wp_enqueue_script( 'foo' );
wp_enqueue_script( 'bar' );
wp_enqueue_script( 'baz' );
```

So the `async` strategy should be making use of `has_only_deferrable_dependents` in a similar way as the `defer` strategy is below.

Reply:
Good point about execution order and async scripts. To clarify, a script can only be `async` if it's dependencies are also `async`, right? will execution order still remain intact if an `async` script's dependency is `deferred`?

Reply:
Actually, I think that if a script is `async` that it doesn't matter what strategy its dependencies has. An `async` script is "evaluated as soon as it is available" which means its timing is not entirely predictable:

<blockquote cite="https://html.spec.whatwg.org/multipage/scripting.html#attr-script-async:~:text=For%20classic%20scripts%2C%20if,effect%20on%20module%20scripts.)"><a href="https://html.spec.whatwg.org/multipage/scripting.html#attr-script-async:~:text=For%20classic%20scripts%2C%20if,effect%20on%20module%20scripts.)"><img src="https://user-images.githubusercontent.com/134745/235752843-7104a558-d62c-4978-9365-95158626318c.png" alt="Cited text from HTML spec"></a></blockquote>

Therefore, I think that if it has dependencies on other scripts on the page, then it must by definition be resilient to being able to be evaluated before/after the dependencies. Otherwise, if an async script absolutely _must_ be assured that its dependencies have been evaluated, then those dependencies must be `blocking` and _not_ be `async` (and not `defer` either). 

Also, if there is a `blocking` script that has an `async` script as a dependency, then this could indicate a case where where the dependent script is no longer eligible for `async` and must instead be `blocking`. Nevertheless, this wouldn't necessarily need to be the case either. If you have an `async` script as a dependency, there are various ways for you to listen for it to be loaded. I put together a demo with the methods that came to mind: https://async-script-load-listeners.glitch.me/

As I think has been discussed already among the Performance team, using `async` scripts is generally not ideal since the loading order is not predictable. Therefore, I expect the strategy would be used infrequently. Nevertheless, when a script truly is purely asynchronous by design (a minority of the time), its dependents would need to know this and would need to implement a method for listening for when it has been loaded. Therefore, the strategy of the dependents should not matter given the unpredictable loading order of the asynchronous script.

Reply:
I would suggest we continue this conversation in a separate location from this line-specific comment thread, as it's a more holistic conversation on how `async` scripts work in terms of dependencies and whether we should include that in the scope of this ticket or in a follow-up ticket or not at all - I guess those are our options.

All I'm saying that this is a higher level discussion that preferably we should have had in the design stage of this work, however of course I acknowledge that concerns can always come up later. I just don't want to derail the code review process itself with this separate conversation :)

FWIW, the current implementation is limited in regards to what you're outlining, but it also doesn't prevent support for that to be added later, so IMO the discussion here should be decoupled from this code review.

Reply:
Happy to move the discussion wherever you like!

Reply:
I think it remains important because if a set of scripts are all purely asynchronous, and there is a dependency tree for these async scripts, if `WP_Scripts` doesn't allow them to have `deps` (or else it makes them all blocking) then a theme/plugin would have to manually keep track of all the dependencies and call `wp_enqueue_script()` for each script in the separately-maintained dependency tree.